lib/patron.rb

```diff
@@ -69,21 +83,26 @@ def self.get_ids_by_rewards(rewards)
         rewards.map { |id| reward_users[id] }.compact.flatten.uniq
       end
 
-      def self.local_users
-        @local_users ||= {}
+      def self.get_local_users
+        users = User.joins(:_custom_fields).where(user_custom_fields: { name: 'patreon_id' }).uniq
+        linked_patron_ids = UserCustomField.where(name: 'patreon_id').where("value IS NOT NULL").pluck(:value)
+
+        oauth_users = Oauth2UserInfo.includes(:user).where(provider: "patreon")
+        oauth_users = oauth_users.where("uid NOT IN (?)", linked_patron_ids) if linked_patron_ids.present?
 
-        @local_users[RailsMultisite::ConnectionManagement.current_db] ||= begin
-          users = Patron.all.map do |p|
-            patreon_id = p[0]
-            patreon_email = p[1]['email']
+        users += oauth_users.map do |o|
+          linked_patron_ids << o.uid
+          update_local_user(o.user, o.uid)
+        end
 
-            user = ::Oauth2UserInfo.find_by(provider: "patreon", uid: patreon_id).try(:user) || ::User.find_by_email(patreon_email)
-            update_local_user(user, patreon_id)
+        emails = all.reject { |p| linked_patron_ids.include?(p[0]) }.map { |p| { p[1]["email"] => p[0] } }.reduce({}, :merge)
```

Comment:
I feel like the data structure that we chose here is wrong judging from how the fact that it takes a `reject`, `map` and `reduce` to filter for the right records.

Reply:
I agree. The data structure we using is little hard to get all emails excluding some ids. I will try to optimize the line for now. Is it okay?

Yes. In future the data structure needs to be changed.

Reply:
I personally feel like we should fix it properly once and for all. Delaying it is just going to cause more code that relies on the current data structure to be added.

Reply:
Okay I will change the data structure in this PR itself. In the meantime to fix the issue I will add another small PR to purge the cache.

Reply:
Maybe @xfalcox can review since I'm going to be away for 2 weeks and he is more familiar with the plugin.

Reply:
Okay. I added that PR https://github.com/discourse/discourse-patreon/pull/25

Reply:
Let me review this over the weekend, I agree we need to improve readability of this method.

Reply:
I changed the data structure little bit. Now the above confusing line removed.